### PR TITLE
Rebase pull request #134 by @joehybird on the latest master

### PR DESCRIPTION
**Pull request by DragoonAethis**
This rebases the Django 2.2 compatibility fixes by @joehybird (PR #134 that is) on the current master, porting all the test changes and Python compatibility cleanups.

I'm 99% sure the tests will fail after the first run, but let's see what I still need to fix :)

### CODE REVIEW

**Comment by richardbarran** [comment]
Hi @DragoonAethis I'm trying to catch up with activity on this project and the various PRs - it looks like your PR is the most complete and up-to-date. I'd like to help out with reviewing - could you take a look at the test that's failing in CI and see if you can quickly fix it?

**Comment by DragoonAethis** [comment]
@richardbarran The only remaining failing tests are on PyPy due to broken dependencies (mysqlclient complaining about `EnvironmentError: mysql_config not found` - fixing this requires installing MySQL/MariaDB client libraries on the host system) - to be honest this is the first time I've touched Travis or CI-testing Python libs and I'm not sure how to fix that. If someone could help over with this, I'd be grateful. Until then, all non-pypy appear to work fine.

**Comment by richardbarran** [comment]
I'm thinking that Pypy support is a "nice to have" - very few django projects out there seem to actively support it (I checked Wagtail, django-cms, Django Rest Framework).
If we just comment out that bit for now, and move forward - if someone is really keen on pypy support then they can help out with the PR?

**Comment by DragoonAethis** [comment]
Removed PyPy from CI - the last test still ran with tox.ini from the previous commit, but it looks ready to merge.

**Comment by richardbarran** [comment]
That's great - I'll pull your branch and try it out on my machine. Give me a day or two to find the time to test it out, and to get back to you.

File: setup.py [hunk 1]
```diff
@@ -17,58 +17,35 @@ def find_version(*file_paths):
...
-try:
-    bytes
-except NameError:
-    bytes = str
-
+def get_version(package):
+    """
+    Return package version as listed in `__version__` in `init.py`.
+    """
+    matches = re.search("__version__[ ]+=[ ]+['\"](?P<version>[^'\"]+)['\"]",
```
Comment (joehybird) [inline setup.py]: Hum... Maybe you should use a raw string here : `r"__version__[\s]+=[\s]+['\"](?P<version>[^'\"]+)['\"]"`

File: .travis.yml [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
+cache: pip
+dist: xenial
 sudo: false
+
+services:
+  - postgresql
+  - mysql
+
+python:
+  - pypy
```
Comment (yakky) [inline .travis.yml]: pypy still declared in travis configuration, maybe we would want to remove this or maybe add to allowed_failures to track pypy support progresses

File: tox.ini [hunk 3]
```diff
@@ -1,28 +1,49 @@
...
+    111: Django>=1.11,<2.0
+    20: Django>=2.0,<2.1
+    21: Django>=2.1,<2.2
+    22: Django>=2.2,<3.0
 setenv =
+    PYTHONPATH = {toxinidir}:{toxinidir}/test_project
     DJANGO_SETTINGS_MODULE = test_project.settings
-    postgres: DJANGO_SETTINGS_MODULE = test_project.postgres_settings
     mysql: DJANGO_SETTINGS_MODULE = test_project.mysql_settings
+    postgresql: DJANGO_SETTINGS_MODULE = test_project.postgres_settings
```
Comment (joehybird) [inline tox.ini]: settings path are declared in commands, so DJANGO_SETTINGS_MODULE is not really useful here

File: .travis.yml (hunk 2 above)
Comment (joehybird) [reply .travis.yml]: This is more an issue with mysql/postgresql dependencies in pypy than the code itself, so i'm agree with @DragoonAethis who want remove it.

**Comment by joehybird** [comment]
Great Job @DragoonAethis :+1:

File: setup.py (hunk 1 above)
Comment (DragoonAethis) [reply setup.py]: Okay, replaced

File: .travis.yml (hunk 2 above)
Comment (DragoonAethis) [reply .travis.yml]: Missed that one when searching for pypy files, will remove

File: tox.ini (hunk 3 above)
Comment (DragoonAethis) [reply tox.ini]: Removed

**Comment by DragoonAethis** [comment]
@joehybird thank you for the port, I just rebased it :P

**Comment by yakky** [comment]
@DragoonAethis @joehybird thanks for your great work! :tada:

**Comment by richardbarran** [comment]
@DragoonAethis @joehybird I've tried this branch on django-photologue - a project that depends on sorted-m2m - with Django 2.2 and it works perfectly.
Excellent work, I'm looking forward to a merge and a release!

**Comment by yakky** [comment]
@richardbarran do you need any help with this and triaging / checking the other related PR?

**Comment by richardbarran** [comment]
@yakky which "other related PR" are you referring to? Do you mean the other PRs that contain code that has been superseded by this PR?

**Comment by richardbarran** [comment]
So, this PR looks good - are we ok to merge? I'll leave this comment up for a few days, and if no one objects then I will go ahead and merge the PR.

**Comment by yakky** [comment]
> @yakky which "other related PR" are you referring to? Do you mean the other PRs that contain code that has been superseded by this PR?

exactly

File: .travis.yml [hunk 4]
```diff
@@ -1,21 +1,31 @@
...
+dist: xenial
 sudo: false
+
+services:
+  - postgresql
+  - mysql
+
+python:
+  - 2.7
+  - 3.4
```
Comment (auvipy) [inline .travis.yml]: 3.4 can be dropped

File: setup.py [hunk 5]
```diff
@@ -87,6 +67,7 @@ def split(self, *args, **kw):
         'Programming Language :: Python :: 3.4',
```
Comment (auvipy) [inline setup.py]: 3.4 should be removed

File: tox.ini [hunk 6]
```diff
@@ -1,28 +1,46 @@
...
 
 [testenv]
-commands = coverage run runtests.py
 deps =
-    111: Django>=1.11a1,<1.11.99
-    postgres: psycopg2
-    mysql: mysqlclient
-    coverage==4.5.3
+    mock<2.0
+    py{27,34}: mysqlclient<1.4
```
Comment (auvipy) [inline tox.ini]: py34 is already EOL

**Comment by DragoonAethis** [comment]
@auvipy I'm cool with dropping 3.4 support (tbh I care primarily about 3.7 at this point, with some lightweight testing on 3.6 and earlier), but before I do that - @richardbarran @joehybird @yakky are you okay with this? Django 1.11 still supports 3.4, 2.0 did but is itself no longer supported, 2.1 supports 3.5+. But then it also doesn't look like too much testing is needed to maintain 3.4 support at this point, so it could be removed along with the next(ish) release, since 1.11 is getting EOLed next April.

**Comment by yakky** [comment]
I'm fine with removing it, as long as it does not create excessive workload for you @DragoonAethis 
As the next will only be a a cleanup / compatibility release, anyone needing python 3.4 can use previous version

**Comment by DragoonAethis** [comment]
Not a problem, one more +1 for deletion and I'll nuke it.

**Comment by richardbarran** [comment]
+1 from me to dropping Python 3.4 support.

**Comment by DragoonAethis** [comment]
Alright, removed. It doesn't look like there was anything else to clean up except test/setup.py entries, so it's good to go.

**Comment by richardbarran** [comment]
> Alright, removed. It doesn't look like there was anything else to clean up except test/setup.py entries, so it's good to go.

Ok, let's merge!

**Pull request merged**